src/main/java/io/supertokens/signingkeys/SigningKeys.java

```diff
@@ -0,0 +1,275 @@
+    }
+
+    public JWTSigningKeyInfo getSigningKeyById(String kid) throws StorageQueryException, StorageTransactionLogicException {
+        return this.getAllKeys().stream()
+                .filter(jwtSigningKeyInfo -> Objects.equals(jwtSigningKeyInfo.keyId, kid)).findAny()
+                .orElse(null);
+    }
+
+    public List<JWTSigningKeyInfo> getAllKeys()
+            throws StorageQueryException, StorageTransactionLogicException {
+
+        return Stream.concat(
+            getDynamicKeys().stream().map(SigningKeys::getJWTSigningKeyInfoFromKeyInfo),
+            getStaticKeys().stream()
+        ).collect(Collectors.toList());
+    }
+
+    public List<KeyInfo> getDynamicKeys() throws StorageQueryException, StorageTransactionLogicException {
+        CoreConfig config = Config.getConfig(main);
+
+        if (this.dynamicKeys == null) {
+            this.dynamicKeys = AccessTokenSigningKey.getInstance(main).getOrCreateAndGetSigningKeys();
+        }
+
+        List<KeyInfo> res = this.dynamicKeys.stream().filter(k -> k.expiryTime >= System.currentTimeMillis()).collect(Collectors.toList());
+
+        if (res.size() == 0 || System.currentTimeMillis() > res.get(0).createdAtTime
+                + config.getAccessTokenDynamicSigningKeyUpdateInterval()) {
+            this.dynamicKeys = null;
+            return getDynamicKeys();
+        }
+
+        return res;
+    }
+
+    public List<JWTSigningKeyInfo> getStaticKeys() throws StorageQueryException, StorageTransactionLogicException {
+        if (this.staticKeys == null) {
+            this.staticKeys = JWTSigningKey.getInstance(main).getAllSigningKeys();
+        }
+
+        return this.staticKeys;
+    }
+    public KeyInfo getLatestIssuedDynamicKey() throws StorageQueryException, StorageTransactionLogicException {
+        return getDynamicKeys().get(0);
+    }
+
+    public synchronized long getDynamicSigningKeyExpiryTime() throws StorageQueryException, StorageTransactionLogicException {
+        long createdAtTime = getLatestIssuedDynamicKey().createdAtTime;
+        return createdAtTime + Config.getConfig(main).getAccessTokenDynamicSigningKeyUpdateInterval();
+    }
+
+    public synchronized void removeKeyFromMemoryIfItHasNotChanged(List<JWTSigningKeyInfo> oldKeyInfo)
+            throws StorageQueryException, StorageTransactionLogicException {
+        // we cannot use read write locks for keyInfo because in getKey, we would
+        // have to upgrade from the readLock to a
+        // writeLock - which is not possible:
+        // https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html
+
+        if (this.dynamicKeys != null && this.staticKeys != null &&
+                // we want to refresh if all keys we are storing were known before
+            this.getAllKeys().stream().allMatch(storedKey -> oldKeyInfo.stream().anyMatch(oldKey -> Objects.equals(oldKey.keyId, storedKey.keyId)))) {
```

Comment:
are you sure this is the correct way to check? Does this account for different sizes in the lengths of the list where one is a full subset of the other?

Reply:
It does: if the new getAllKeys is a subset of the old key list, we still need to hit the DB to ensure.